src/voter/mod.rs

```diff
@@ -89,10 +89,20 @@ pub trait Environment<H: Eq, N: BlockNumberOps>: Chain<H, N> {
 	fn proposed(&self, round: u64, propose: PrimaryPropose<H, N>) -> Result<(), Self::Error>;
 
 	/// Note that we have prevoted in the given round.
-	fn prevoted(&self, round: u64, prevote: Prevote<H, N>) -> Result<(), Self::Error>;
+	fn prevoted(
+		&self,
+		round: u64,
+		prevote: Prevote<H, N>,
+		votes: &HistoricalVotes<H, N, Self::Signature, Self::Id>,
+	) -> Result<(), Self::Error>;
 
 	/// Note that we have precommitted in the given round.
-	fn precommitted(&self, round: u64, precommit: Precommit<H, N>) -> Result<(), Self::Error>;
+	fn precommitted(
+		&self,
+		round: u64,
+		precommit: Precommit<H, N>,
+		votes: &HistoricalVotes<H, N, Self::Signature, Self::Id>,
```

Comment:
This is passed along when prevoting/precommitting because we might need to answer a challenge in the current round?

Reply:
I thought it could end up in an inconsistent state if historical votes are not saved at prevote/precommit. I'm not sure now, but I guess it doesn't hurt.